lxml-stubs/etree.pyi

```diff
@@ -56,9 +56,17 @@ _XPathObject = Union[
 _ListAnyStr = Union[List[str], List[bytes]]
 _DictAnyStr = Union[Dict[str, str], Dict[bytes, bytes]]
 _Dict_Tuple2AnyStr_Any = Union[Dict[Tuple[str, str], Any], Tuple[bytes, bytes], Any]
-_NSMap = Union[Dict[Union[bytes, None], bytes], Dict[Union[str, None], str]]
 _xpath = Union["XPath", _AnyStr]
-_OptionalNamespace = Optional[Mapping[str, Any]]
+
+# See https://github.com/python/typing/pull/273
+# Due to Mapping having invariant key types, Mapping[Union[A, B], ...]
+# would fail to validate against either Mapping[A, ...] or Mapping[B, ...]
+# Try to settle for simpler solution, encouraging use of empty string ('')
+# as default namespace prefix. If too many people complain, it can be
+# back-paddled as Mapping[Any, ...]
+_NSMapArg = Optional[Mapping[str, str]]
+_NonDefaultNSMapArg = Optional[Mapping[str, str]]  # empty prefix disallowed
```

Comment:
I think the `Optional[]` marker should stay with the argument, not its type. It's the argument that is optional.

Reply:
Hmph, as expected, the naming creates some misunderstanding. This is my intention:

- `NSMapArg`: default namespace is allowed, like `{'': 'http://default.ns/', 'prefix1': 'ns1'}`
- `NonDefaultNSMapArg`: `{'prefix1': 'ns1'}`, default namespace is disallowed.

`Optional` are present in both type aliases because all namespaces related argument is `None` if not supplied, it has nothing to do with ns prefixes.

If `Optional` is added to the mapping key type as `Mapping[Optional[str], str]`, then `{str: str}` will actually  _fail_ to validate, as I have described in #31. `{Optional[str]: str}` and `{str: str}` are incompatible to each other under current typing system, and there is no workaround except relaxing to `Any`.

Reply:
I understand that. However, the fact that the argument is optional is part of the function signature, not part of the type of an argument. It's a specific function that allows it to be optional. Whether that applies to all functions or not isn't really relevant.

Reply:
Guess I sort of understand what you mean now. Dropping `Optional` in the alias definition itself and apply `Optional` on each function signature instead? Can update the PR tomorrow.